### ISSUE TITLE
Custom Functions that accept Entity arguments can take a long time and fail to calculate when there are many on a worksheet/book

### ISSUE THREAD

**Issue by demyte** [OfficeDev/office-js#4654]
We are using streaming functions in our addin where users can pass the results of some of our other functions that produce Entity results in as arguments.

We are seeing an issue on larger workbooks where when the workbook calculates (e.g. when it is first opened) and our functions fire, any of our functions that have entity cells passed in as arguments take an excessively long time to actually hit our function code.

In the extreme example, this delay can last for 70+ seconds and then fail before we ever have our function code triggered and the following error is showing in the devtools console
 
![image](https://github.com/OfficeDev/office-js/assets/48166/a9c9e093-513f-4e72-bccd-ef1a89ceafdc)

The top of this file is as follows if it helps to understand the context of where it is going wrong:

![image](https://github.com/OfficeDev/office-js/assets/48166/1e809970-9b66-4f81-947b-74884f836cf3)

I have attached a trace file [Trace-20240709T165130.zip](https://github.com/user-attachments/files/16139635/Trace-20240709T165130.zip) that I started and then triggered a workbook calculation about 2 seconds after.  You can see that not much goes on until after 80seconds when the error above is displayed

UPDATE: See sample repo below that shows it happening

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC Desktop
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: Microsoft® Excel® for Microsoft 365 MSO (Version 2406 Build 16.0.17726.20078) 64-bit 
* Operating System: Windows 11 x64
* Browser (if using Office on the web): 
 
## Expected behavior

Expectation is that these formula being passed an entity get through the Excel side of the engine to our side with the same as or close to the same performance as a standard formula with normal arguments.

## Current behavior

As above... it can be exceptionally slow when there are many formula and can lead to a browser error and even Excel locking up and crashing

## Steps to reproduce

1. Using the sample
2. Build and run it
3. Load the example excel workbook from the repo
4. Calculate it

## Link to live example(s)
This sample is a clean `yo office` generated shared runtime project with only the functions updated:

- https://github.com/demyte/officejs-failing-fn-sharedruntime

Note: These two functions and workbook example work in a ScriptLab sample.

**Comment by demyte** [comment]
UPDATE:

I have a breaking sample located here: https://github.com/demyte/officejs-failing-fn-sharedruntime

There is a [README](https://github.com/demyte/officejs-failing-fn-sharedruntime/blob/main/README.md) with instructions.

I note that when I use the same functions in a script lab sample, it works, so I assume its an issue with the SharedRuntime.

**Comment by shanshanzheng-dev** [comment]
Hi @demyte Thanks for reporting this issue. We'll be looking into this issue and will report back if we have a suggestion for you, Thanks.

**Comment by demyte** [comment]
@shanshanzheng-dev 

I just wanted to also note that I scaffold a non-shared runtime addin and popped the functions in there and it works just like in script lab and successfully fires and resolves all the functions in the test workbook.

Also, the test workbook doesnt have to be as big as I have there for the failure to occur, I just pushed it out large to test shared v non-shared runtimes and what worked / didn't.

**Comment by shanshanzheng-dev** [comment]
Hi @demyte Thanks for providing these information. That helps to triage this issue better. So this issue only hit on shareRuntime add-in. right?
As you mention, I try to repro your issue using your test add-in, when I change a sheet1:A1 number, it occur "NAME?" error.
![image](https://github.com/OfficeDev/office-js/assets/125014206/3ef11c2f-2b68-4225-9d79-8523c9b8178f)

**Comment by demyte** [comment]
@shanshanzheng-dev 

Oops - thats my testing file from script lab - its the same file, just update with a find and replace:

FIND: `_xldudf_SCRIPTLAB_STREAMINGFUNCTION_`
REPLACE: `CONTOSO.`
WITHIN:  Workbook

e.g.
![image](https://github.com/OfficeDev/office-js/assets/48166/cc0d95ab-19ad-4d99-9cd0-22e3b4d55121)

I'll update the repo in a little bit.

**Comment by demyte** [comment]
@shanshanzheng-dev the repo has been updated with a new test file

**Comment by shanshanzheng-dev** [comment]
Hi @demyte Thanks. Seem FnMain() doesn't work, FnEntity() works good. If my understanding is right, seem there is some delay to re-calculates when I change sheet1:A1. This is your concern, right? We'll be looking into this issue.
![image](https://github.com/OfficeDev/office-js/assets/125014206/8ae5e685-f670-40d4-aa89-deae8f6ee01e)
![image](https://github.com/OfficeDev/office-js/assets/125014206/3f9769a3-5d4b-4854-a0c2-a40ad16fe179)

**Comment by demyte** [comment]
@shanshanzheng-dev 

Yes, that is correct, fnMain never actually calls the function code when there is a critical mass of that formula on the sheet/workbook.  

I can't say what the critical mass is.. but whatever it is, it looks like a json representation of it all is getting truncated somewhere in the broker between your Excel core function code and the sharedruntime host before it gets to call our function callback. (thats a massive guess - i have no idea about whats happening on the Excel side of the fence :) )

If you have your browser dev tools open, you will see the error in my original post after somewhere 60+ seconds from calculating - this can depend on your machine spec it seems - just wait a few minutes and it will pop up :)

![image](https://github.com/OfficeDev/office-js/assets/48166/ce123359-784c-487f-843a-a260014658f7)

**Comment by shanshanzheng-dev** [comment]
Hi @demyte Thanks for confirming. I can see this error message. We'll be looking into this issue. Thanks for your time.

**Comment by demyte** [comment]
@shanshanzheng-dev - is there any update on this?

This is a very common issue for our customers.

We can't even handle this error and let users know as its happening in the Excel runtime side, so for them, their workbook just looks like it is calculating forever and then they blame our addin.

**Comment by shanshanzheng-dev** [comment]
Hi @demyte Sorry for low response. We're investigating the root cause. It has been put on our backlog <bug #9154685>. We unfortunately have no timelines to share at this point.

**Comment by demyte** [comment]
Hi @shanshanzheng-dev,

I am looking to get an update on this issue as it affects our users badly when they have workbooks of a certain size / complexity or greater (they are not massive sheets, the issue starts appearing noticably when we start to hit ~ 1500-2000 cells and gets worse from there).

This and the other two issues I currently have open (#4679 & #4487) correlate to over 75% of all the support requests we get from our users.

**Comment by Dhoot** [comment]
hello @demyte we apologize for the lack of a substantive update.

We will investigate based on the information provided. To help us further, a few quick follow-ups:

1. Is this still reproducible on your current version of Excel and Office JS? If yes, please share your platform (Windows/Mac/Web) and Office build number.
2. Are you still actively blocked, or have you found any workaround that is acceptable in production? (We noted throttling update rates was explored but didn't meet real-time requirements.)

Has the behavior changed with any newer Office versions you may have tested since your last comment?

We will share an update on our findings as soon as we have something concrete to report.

**Comment by jpickles1981** [comment]
Hi @Dhoot , I'd like to report the same issue in an add-in I have developed.  I have some custom functions which accept Entity arguments as input.  When my workbook contains many of these functions, e.g. ~1800 in my example workbook and they are invoked simultaneously, I see #BUSY! in the cell indefinitely and I see the same error in the console log as the original post.  if I invoke these functions in batches, e.g. 100 at a time, then I can complete all 1800.  However, if I try to invoke all 1800 at once, then they never complete and all of them say #BUSY!.

I see it's been two years since this bug was first reported, which is disappointing to see that it hasn't been fixed yet.  This is a signficiant problem for me, as it's extremely limiting having to tell users to avoid >1000 custom functions in their workbook.

This is what I would see in my workbook.  You can see I am passing in entities in column F into the custom functions in column G.

<img width="1273" height="699" alt="Image" src="https://github.com/user-attachments/assets/f04fe488-9edb-46be-b3a8-e4a172a86682" />

I am using 
- Excel desktop
- 64-bit
- Version 2605 (Build 20026.20166)
- Shared runtime add-in

**Comment by demyte** [comment]
@Dhoot 

Yes this is still an issue, nothing has changed.  If you spin up the sample repo it will show you.

**There is no satisfactory workaround for us.**

We have had to implement a feature into our product that warns users that they are getting sheets that are to large for performant calculation and help them through modifying their sheets to be as performant as possible.


### Environment
**Platform [PC desktop, Mac, iOS, Office on the web]:** All (don't even think about trying this on Web tho.. it is orders of magnitude worse)
**Host [Excel, Word, PowerPoint, etc.]:** Excel - Microsoft® Excel® for Microsoft 365 MSO (Version 2605 Build 16.0.20026.20182) 64-bit 
**Operating System:** Windows 11 x64
**Browser (if using Office on the web):** Edge/Chrome

**Comment by Dhoot** [comment]
Thanks @demyte we are tracking WI #9154685 (internal) we will communicate with you for further clarification to address it

**Comment by jgupta8450** [comment]
Thank you for your patience.

After investigation, this is **not a product bug** but an **existing platform constraint**: Excel limits custom-function payloads to **5 MB**. When a custom function takes rich Entity values as arguments, they're serialized into one payload; at scale that exceeds 5 MB and the function fails to run. (The workbook file is compressed, so its size isn't the measure — the serialized payload is several times larger.)

This is **by design**, so the change is a **feature request, not a bug fix**. After reviewing the reported behaviour, we have determined that this is a feature request rather than a product bug. If you would like us to consider this enhancement for a future release, please submit your suggestion to the Microsoft Tech Community Ideas Forum or upvote an existing idea. This allows other developers with similar requirements to provide feedback and helps our product teams assess community interest.

🔗 Ideas Forum: [https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/idb-p/Microsoft365Developer…](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/idb-p/Microsoft365Developer…)

📄 Documented limit: [https://learn.microsoft.com/en-us/office/dev/add-ins/concepts/resource-limits-and-performance-optimization](https://learn.microsoft.com/en-us/office/dev/add-ins/concepts/resource-limits-and-performance-optimization)

**Workarounds:**

- Fewer FNMAIN calls / sheets to stay under the threshold.
- Reduce per-entity payload (fewer properties/metadata).
- Pass a key instead of a full entity; have FNMAIN look it up.
- Calculate fewer entity cells at once (e.g. manual recalc, one sheet at a time) to keep each batch small.

Thank you for your understanding and support!

**Comment by demyte** [comment]
In my testing, the performance issue starts occuring a long way before we get anywhere close to the 5MB limit.  This performance issue is really what needs to be looked at as it affects users without any error being able to be bubbled back to them. i.e. if we hit the 5MB limit we would get an error we could handle.

So, i can't really see how this is a feature request not a bug when we are still well within the 5mb limit.

As I mentioned previously, we do have a warning that shows to users that they are starting to hit (our artificially set) limits and we offer a feature that does the replace `<entitycell>` with `<entitycell>.key` as a way of mitigating things to a point.

The workarounds of "Calculate fewer entity cells at once" and "Fewer FNMAIN cells" are not at all achievable with auto-calc on.

**Comment by jgupta8450** [comment]
Thanks for the response. When a custom function consumes Entity arguments, Excel's calc engine serializes the invocation into a single JSON payload and sends it into the shared runtime. There's a native inbound transport limit (~4 MB); once the payload crosses it, the message arrives truncated,  JSON.parse  fails ("Unexpected end of JSON input"), and FNMAIN never runs. That's why you hit it well before 5 MB, and why it's shared‑runtime‑only.

We will work on the enhancement (no timeline yet): raising the actual ceiling needs inbound chunking + reassembly

We agree that "fewer cells / fewer FNMAIN" suggestions aren't viable with auto‑calc on. The one real mitigation or workaround for now is shrinking per‑call payload — pass a key instead of a full entity and look it up inside FNMAIN.

To help pin down the threshold:

1. How much data triggers it — roughly how many FNMAIN cells / entity args (or payload size) before it fails, and is there a level that's reliably fine for you? It will help us correlate to the ~4 MB limit.

Thanks again. we're tracking this internally and will keep this issue updated.

**Comment by jpickles1981** [comment]
This should not be closed, as it remains unresolved.

**Comment by jpickles1981** [comment]
Is there any update on this?

**Comment by ayush-meghwani-28** [comment]
Hello @jpickles1981 Apologies for not providing regular updates here.

The internal item ID for this issue is `9154685`. I’m currently investigating it and have identified the main concern and issue details. I’ve also created a self-repro add-in and am now examining the root cause.

I’ll provide an update once I’ve identified the next best steps. If this is a high-priority issue for you, please let me know so we can triage it accordingly. Thanks for your patience!

**Comment by jpickles1981** [comment]
Yes, its high priority

**Comment by demyte** [comment]
@ayush-meghwani-28 this is still really important to us to have this working effectively, but based on previous comments by the MS team, it seemed that it was thrown into the `too hard basket`.

I beleive that there is more than enough info in my initial and subsequent submissions to go on.
